File: claude/.claude/skills/shortcut/scripts/epics.py
#!/usr/bin/env python3
"""Shortcut Epics - Operations for managing epics."""

import sys
import json
import argparse
from typing import Dict, List, Optional
from shortcut_client import ShortcutClient, format_epic


def get_epic(epic_id: int) -> Dict:
    """Get a single epic by ID."""
    client = ShortcutClient()
    epic = client.get(f"epics/{epic_id}")
    return format_epic(epic)


def list_epics() -> List[Dict]:
    """List all epics."""
    client = ShortcutClient()
    epics = client.get("epics")
    return [format_epic(epic) for epic in epics]


def search_epics(
    query: Optional[str] = None, state: Optional[str] = None
) -> List[Dict]:
    """Search epics."""
    client = ShortcutClient()
    params = {}
    if query:
        params["query"] = query
    if state:
        params["state"] = state

    epics = client.get("epics", params=params)
    return [format_epic(epic) for epic in epics]


def create_epic(
    name: str,
    description: Optional[str] = None,
    state: str = "to do",
    owner_ids: Optional[List[str]] = None,
    milestone_id: Optional[int] = None,
) -> Dict:
    """Create a new epic."""
    client = ShortcutClient()

    epic_data = {"name": name, "state": state}
    if description:
        epic_data["description"] = description
    if owner_ids:
        epic_data["owner_ids"] = owner_ids
    if milestone_id:
        epic_data["milestone_id"] = milestone_id

    epic = client.post("epics", epic_data)
    return format_epic(epic)


def update_epic(
    epic_id: int,
    name: Optional[str] = None,
    description: Optional[str] = None,
    state: Optional[str] = None,
    owner_ids: Optional[List[str]] = None,
    archived: Optional[bool] = None,
) -> Dict:
    """Update an existing epic."""
    client = ShortcutClient()

    update_data = {}
    if name is not None:
        update_data["name"] = name
    if description is not None:
        update_data["description"] = description
    if state is not None:
        update_data["state"] = state
    if owner_ids is not None:
        update_data["owner_ids"] = owner_ids
    if archived is not None:
        update_data["archived"] = archived

    epic = client.put(f"epics/{epic_id}", update_data)
    return format_epic(epic)


def delete_epic(epic_id: int) -> Dict:
    """Delete an epic."""
    client = ShortcutClient()
    client.delete(f"epics/{epic_id}")
    return {"success": True, "message": f"Epic {epic_id} deleted"}


def main():
    parser = argparse.ArgumentParser(description="Shortcut Epics Operations")
    subparsers = parser.add_subparsers(dest="command")

    get_parser = subparsers.add_parser("get")
    get_parser.add_argument("epic_id", type=int)

    subparsers.add_parser("list")

    search_parser = subparsers.add_parser("search")
    search_parser.add_argument("--query")
    search_parser.add_argument("--state")

    create_parser = subparsers.add_parser("create")
    create_parser.add_argument("name")
    create_parser.add_argument("--description")
    create_parser.add_argument("--state", default="to do")
    create_parser.add_argument("--owner-ids", nargs="+")

    update_parser = subparsers.add_parser("update")
    update_parser.add_argument("epic_id", type=int)
    update_parser.add_argument("--name")
    update_parser.add_argument("--description")
    update_parser.add_argument("--state")
    update_parser.add_argument("--archived", action="store_true")

    delete_parser = subparsers.add_parser("delete")
    delete_parser.add_argument("epic_id", type=int)

    args = parser.parse_args()

    try:
        if args.command == "get":
            result = get_epic(args.epic_id)
        elif args.command == "list":
            result = list_epics()
        elif args.command == "search":
            result = search_epics(args.query, args.state)
        elif args.command == "create":
            result = create_epic(
                args.name, args.description, args.state, args.owner_ids
            )
        elif args.command == "update":
            result = update_epic(
                args.epic_id,
                args.name,
                args.description,
                args.state,
                args.owner_ids if hasattr(args, "owner_ids") else None,
                args.archived if hasattr(args, "archived") else None,
            )
        elif args.command == "delete":
            result = delete_epic(args.epic_id)
        else:
            parser.print_help()
            sys.exit(1)

        print(json.dumps(result, indent=2))
    except Exception as e:
        print(json.dumps({"error": str(e)}, indent=2), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: claude/.claude/skills/shortcut/src/entities/epics.ts
import type { Command } from "commander";
import type { ApiEpic, FormattedEpic, DeleteResult } from "../types.js";
import { ShortcutClient, formatEpic, output, outputError } from "../client.js";

export async function getEpic(epicId: number): Promise<FormattedEpic> {
  const client = new ShortcutClient();
  const epic = await client.get<ApiEpic>(`epics/${epicId}`);
  return formatEpic(epic);
}

export async function listEpics(): Promise<FormattedEpic[]> {
  const client = new ShortcutClient();
  const epics = await client.get<ApiEpic[]>("epics");
  return epics.map(formatEpic);
}

export async function searchEpics(
  query?: string,
  state?: string,
): Promise<FormattedEpic[]> {
  const client = new ShortcutClient();
  const params: Record<string, string> = {};
  if (query) params.query = query;
  if (state) params.state = state;

  const epics = await client.get<ApiEpic[]>("epics", params);
  return epics.map(formatEpic);
}

export interface CreateEpicOptions {
  name: string;
  description?: string;
  state?: string;
  ownerIds?: string[];
  milestoneId?: number;
}

export async function createEpic(
  options: CreateEpicOptions,
): Promise<FormattedEpic> {
  const client = new ShortcutClient();
  const { name, description, state = "to do", ownerIds, milestoneId } = options;

  const epicData: Record<string, unknown> = { name, state };
  if (description) epicData.description = description;
  if (ownerIds) epicData.owner_ids = ownerIds;
  if (milestoneId) epicData.milestone_id = milestoneId;

  const epic = await client.post<ApiEpic>("epics", epicData);
  return formatEpic(epic);
}

export interface UpdateEpicOptions {
  epicId: number;
  name?: string;
  description?: string;
  state?: string;
  ownerIds?: string[];
  archived?: boolean;
}

export async function updateEpic(
  options: UpdateEpicOptions,
): Promise<FormattedEpic> {
  const client = new ShortcutClient();
  const { epicId, name, description, state, ownerIds, archived } = options;

  const updateData: Record<string, unknown> = {};
  if (name !== undefined) updateData.name = name;
  if (description !== undefined) updateData.description = description;
  if (state !== undefined) updateData.state = state;
  if (ownerIds !== undefined) updateData.owner_ids = ownerIds;
  if (archived !== undefined) updateData.archived = archived;

  const epic = await client.put<ApiEpic>(`epics/${epicId}`, updateData);
  return formatEpic(epic);
}

export async function deleteEpic(epicId: number): Promise<DeleteResult> {
  const client = new ShortcutClient();
  await client.delete(`epics/${epicId}`);
  return { success: true, message: `Epic ${epicId} deleted` };
}

export function registerReadCommands(program: Command): void {
  const epics = program.command("epics").description("Epic operations");

  epics
    .command("get <epic-id>")
    .description("Get an epic by ID")
    .action(async (epicId: string) => {
      try {
        const result = await getEpic(parseInt(epicId, 10));
        output(result);
      } catch (e) {
        outputError(e);
      }
    });

  epics
    .command("list")
    .description("List all epics")
    .action(async () => {
      try {
        const result = await listEpics();
        output(result);
      } catch (e) {
        outputError(e);
      }
    });

  epics
    .command("search")
    .description("Search epics")
    .option("--query <query>", "Search query")
    .option("--state <state>", "Epic state")
    .action(async (opts) => {
      try {
        const result = await searchEpics(opts.query, opts.state);
        output(result);
      } catch (e) {
        outputError(e);
      }
    });
}

export function registerWriteCommands(program: Command): void {
  const epics = program.command("epics").description("Epic operations");

  epics
    .command("create <name>")
    .description("Create a new epic")
    .option("--description <desc>", "Epic description")
    .option("--state <state>", "Epic state", "to do")
    .option("--owner-ids <ids...>", "Owner IDs")
    .option("--milestone-id <id>", "Milestone ID")
    .action(async (name: string, opts) => {
      try {
        const result = await createEpic({
          name,
          description: opts.description,
          state: opts.state,
          ownerIds: opts.ownerIds,
          milestoneId: opts.milestoneId
            ? parseInt(opts.milestoneId, 10)
            : undefined,
        });
        output(result);
      } catch (e) {
        outputError(e);
      }
    });

  epics
    .command("update <epic-id>")
    .description("Update an epic")
    .option("--name <name>", "Updated name")
    .option("--description <desc>", "Updated description")
    .option("--state <state>", "Updated state")
    .option("--owner-ids <ids...>", "Owner IDs")
    .option("--archived", "Archive the epic")
    .action(async (epicId: string, opts) => {
      try {
        const result = await updateEpic({
          epicId: parseInt(epicId, 10),
          name: opts.name,
          description: opts.description,
          state: opts.state,
          ownerIds: opts.ownerIds,
          archived: opts.archived,
        });
        output(result);
      } catch (e) {
        outputError(e);
      }
    });

  epics
    .command("delete <epic-id>")
    .description("Delete an epic")
    .action(async (epicId: string) => {
      try {
        const result = await deleteEpic(parseInt(epicId, 10));
        output(result);
      } catch (e) {
        outputError(e);
      }
    });
}
